### PR TITLE
ForwardedHeaders must be called before UseHsts

### LINKED ISSUE #18816: `UseForwardedHeaders` and middleware order in the code sample
In the [code sample](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1#other-proxy-server-and-load-balancer-scenarios
) for "Other proxy server and load balancer scenarios" `app.UseForwardedHeaders()` is invoked before everything else, including `UseExceptionHandler`/`UseDeveloperExceptionPage`.  While on [this docs](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-nginx?view=aspnetcore-3.1#use-a-reverse-proxy-server) page next is stated:

> As a general rule, Forwarded Headers Middleware should run before other middleware except diagnostics and error handling middleware. 

What is correct approach? Is  `UseExceptionHandler`/`UseDeveloperExceptionPage` error handling middleware or am i missing something?

Thank you.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e548cd16-b06c-5802-a1a7-c627167857ae
* Version Independent ID: c48abb08-bea3-dcff-c067-2edf5dc9b3a6
* Content: [Configure ASP.NET Core to work with proxy servers and load balancers](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1)
* Content Source: [aspnetcore/host-and-deploy/proxy-load-balancer.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/host-and-deploy/proxy-load-balancer.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-hostdeploy**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #18816

Internal review URLs
* [Forwarded Headers Middleware order](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1&branch=pr-en-us-18821#forwarded-headers-middleware-order)
* [apache](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-apache?view=aspnetcore-2.1&branch=pr-en-us-18821#configure-a-proxy-server)
https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-nginx?view=aspnetcore-2.1&branch=pr-en-us-18821#use-a-reverse-proxy-server

### CODE REVIEW

File: aspnetcore/includes/ForwardedHeaders.md [hunk 1]
```diff
@@ -0,0 +1 @@
+As a general rule, Forwarded Headers Middleware should run before other middleware. This ordering ensures that the middleware relying on forwarded headers information can consume the header values for processing. Forwarded Headers Middleware can run after diagnostics and error handling middleware, but it must must be called before `UseHsts`.
```
Comment (Tratcher) [inline aspnetcore/includes/ForwardedHeaders.md:1]: Calling out just UseHsts here doesn't make sense when this paragraph is included in docs that don't show the context for Startup.cs. The line from the apache and nginx docs covers this in a more general way.
"Any component that depends on the scheme, such as authentication, link generation, redirects, and geolocation, must be placed after invoking the Forwarded Headers Middleware."

That could be incorporated in here and then you could give a few examples at the end like UseHsts, UseHttpsRedirection, UseAuthentication, etc..

Comment (Rick-Anderson) [reply aspnetcore/includes/ForwardedHeaders.md:1]: Here's the original paragraph in apache/nginx

>Any component that depends on the scheme, such as authentication, link generation, redirects, and geolocation, must be placed after invoking the Forwarded Headers Middleware. As a general rule, Forwarded Headers Middleware should run before other middleware ***except diagnostics and error handling middleware. *** This ordering ensures that the middleware relying on forwarded headers information can consume the header values for processing.

***except diagnostics and error handling middleware. ***  is what needs to be fixed.

> like UseHsts, UseHttpsRedirection, UseAuthentication, etc..

`UseHttpsRedirection, UseAuthentication` are covered with the new include

As a general rule, Forwarded Headers Middleware should run before other middleware.

I'll change this to say:

~As a general rule,~ Forwarded Headers Middleware should run before other middleware.

That covers everything. I don't see the problem with

Forwarded Headers Middleware **_can_** run after diagnostics and error handling middleware, but it must must be called before `UseHsts`.

`UseHsts` is the only problematic middleware call the templates generate as it's grouped with diagnostics and error handling middleware

**Comment by Rick-Anderson** [comment]
We should probably just go with 
Forwarded Headers Middleware should run before other middleware.

and drop 

Forwarded Headers Middleware can run after diagnostics and error handling middleware, but it must must be called before `UseHsts`.

**Comment by Rick-Anderson** [comment]
**Original** :
Any component that depends on the scheme, such as authentication, link generation, redirects, and geolocation, must be placed after invoking the Forwarded Headers Middleware. As a general rule, Forwarded Headers Middleware should run before other middleware **_except diagnostics and error handling middleware_**. This ordering ensures that the middleware relying on forwarded headers information can consume the header values for processing.

**_Invoke the UseForwardedHeaders method at the top of Startup.Configure before calling other middleware._** Configure the middleware to forward the X-Forwarded-For and X-Forwarded-Proto headers:


**Now:**

Any component that depends on the scheme, such as authentication, link generation, redirects, and geolocation, must be placed after invoking the Forwarded Headers Middleware.

Forwarded Headers Middleware should run before other middleware. This ordering ensures that the middleware relying on forwarded headers information can consume the header values for processing. To run Forwarded Headers Middleware after diagnostics and error handling middleware, see [Forwarded Headers Middleware order](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1&branch=pr-en-us-18821#forwarded-headers-middleware-order).

**_Invoke the UseForwardedHeaders method at the top of Startup.Configure before calling other middleware._** Configure the middleware to forward the X-Forwarded-For and X-Forwarded-Proto headers:

**Pull request merged**